### PR TITLE
Test core functionalities

### PR DESCRIPTION
**Pull request by delta2323**
This PR is implementation of #5 . It includes :
* More unittest for unary and binary operations for `Variable`
* Unittest for `Variable`
* Unittest for `FunctionSet` where `FunctionSet` is nested.
* Unittest for utility of `Optimizer`
* Functionality test for each optimizers

Since issue #5 contains implementation of tests for multiple modules, I would recommend we should split it into more smaller issues.

### CODE REVIEW

**Comment by unnonouno** [comment]
How about splitting this issue into six issues. You can make small issues with cherry-pick easily.

**Comment by delta2323** [comment]
I propose split of this PR into :
* unittest for operation on Variable
  * Add unittest for binary operators where arguments are numpy.ndarray a… …    97e974a
* unittest for Variable 
  * Add unittest for Variable.backward    dcb1f4e
  * Add unittest for Variable.unchain_backward    af1f5b2
* Functionality test for optimizer
  * Add unittest for  1012407
  * Unittest of SGD creates new training data in every epoch    45fe8e9
  * Support unittest of sgd for gpu    d8c13c7
  * Extract linear model with train data generator from TestSGD class    95595ee
  * Add functionality tests using linear model for optimizers other than SGD    d85953a
* unittest for Function set
  * Add unittest for FunctionSet._get_sorted_funcs    f15e122
  * Add unittest for nested function sets    6311685
* unittest for optimizer
  * Add unittest for utility functions of optimizer    688fd5f

So # of resulting smaller PRs will be five. If we can split into smaller parts as @unnonouno suggests, I will examine it.

**Comment by beam2d** [comment]
Great work! I will review it shortly, and decide whether further tests should be included or not into this PR.

**Comment by beam2d** [comment]
I think this PR can stay as is without splitting. I'm reviewing it.
@delta2323 Do you have any plan for further tests? If you want more tests, they should be put into other issues/PRs.

File: tests/functions_tests/test_basic_math.py [hunk 1]
```diff
@@ -157,6 +157,95 @@ def test_rdiv_backward_gpu(self): self.backward_gpu(lambda x, y: y / x)
...
+            value = cuda.to_gpu(value)
+        x = Variable(x_data)
+        y = op(x, value)
+        assert_allclose(op(self.x, self.value), y.data, atol=1e-6, rtol=1e-6)
+
+    def forward_cpu(self, op):
+        self.check_forward(op, self.x, False)
+
+    # 2015/6/15 Kenta OONO <oono@preferred.jp>
+    # Commented tests do not work. See issue #43
```
Comment (beam2d) [inline tests/functions_tests/test_basic_math.py]: Write it as a TODO comment, and do not write date information to comments (it can be extracted from the repository log).

File: tests/functions_tests/test_basic_math.py [hunk 2]
```diff
@@ -157,6 +157,95 @@ def test_rdiv_backward_gpu(self): self.backward_gpu(lambda x, y: y / x)
...
+    # def test_rmul_backward_cpu(self): self.backward_cpu(lambda x, y: y * x)
+    def test_div_backward_cpu(self):  self.backward_cpu(lambda x, y: x / y)
+    # def test_rdiv_backward_cpu(self): self.backward_cpu(lambda x, y: y / x)
+    def test_pow_backward_cpu(self):  self.backward_cpu(lambda x, y: x ** y)
+    # def test_rpow_backward_cpu(self): self.backward_cpu(lambda x, y: y ** x)
+
+    def backward_gpu(self, op):
+        self.check_backward(op, to_gpu(self.x), to_gpu(self.gy), True)
+
+    # 2015/6/15 Kenta OONO <oono@preferred.jp>
```
Comment (beam2d) [inline tests/functions_tests/test_basic_math.py]: Remove date and author information from regular comments. They can be extracted from the repository log.

File: tests/optimizers_tests/linear_model.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            l = Linear(LinearModel.UNIT_NUM, 2)
+        )
+        self.optimizer = optimizer
+        # true parameters
+        self.w      = np.random.uniform(-1, 1, (LinearModel.UNIT_NUM, 1)).astype(np.float32)
+        self.b      = np.random.uniform(-1, 1, (1, )).astype(np.float32)
+
+    def _train_linear_classifier(self, model, optimizer, gpu):
+        def _make_label(x):
+            a = (np.dot(x, self.w) + self.b).reshape((LinearModel.BATCH_SIZE, ))
```
Comment (beam2d) [inline tests/optimizers_tests/linear_model.py]: Use `self.BATCH_SIZE` instead of `LinearModel.BATCH_SIZE`, since class attributes can be overridden by child classes.

File: tests/optimizers_tests/linear_model.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            x_data = np.random.uniform(-1, 1, (batch_size, unit_num)).astype(np.float32)
+            t_data = _make_label(x_data)
+            if gpu:
+                x_data = cuda.to_gpu(x_data)
+                t_data = cuda.to_gpu(t_data)
+            x = Variable(x_data)
+            t = Variable(t_data)
+            return x, t
+
+        for epoch in xrange(LinearModel.EPOCH):
```
Comment (beam2d) [inline tests/optimizers_tests/linear_model.py]: Ditto: `LinearModel` to `self`

File: tests/optimizers_tests/linear_model.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            x, t = _make_dataset(LinearModel.BATCH_SIZE, LinearModel.UNIT_NUM, gpu)
```
Comment (beam2d) [inline tests/optimizers_tests/linear_model.py]: Ditto

File: tests/optimizers_tests/linear_model.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            optimizer.zero_grads()
+            y = model.l(x)
+            loss = softmax_cross_entropy(y, t)
+            loss.backward()
+            optimizer.update()
+
+        x_test, t_test = _make_dataset(LinearModel.BATCH_SIZE, 10, gpu)
```
Comment (beam2d) [inline tests/optimizers_tests/linear_model.py]: ditto

File: tests/test_function_set.py [hunk 7]
```diff
@@ -1,10 +1,63 @@
...
 cuda.init()
 
+class MockFunction(Function):
+    def __init__(self, shape):
+        self.p = np.zeros(shape).astype(np.float32)
+        self.gp = np.ones(shape).astype(np.float32)
+
+    @property
+    def parameter_names(self):
+        return ('p', )
```
Comment (beam2d) [inline tests/test_function_set.py]: It can be a class attribute instead of a property. Replace it by
```
    parameter_names = 'p',
```

File: tests/test_function_set.py [hunk 8]
```diff
@@ -1,10 +1,63 @@
...
+
+    @property
+    def gradient_names(self):
+        return ('gp', )
```
Comment (beam2d) [inline tests/test_function_set.py]: ditto. Replace it by
```
    gradient_names = 'gp',
```

File: tests/test_function_set.py [hunk 9]
```diff
@@ -1,10 +1,63 @@
...
+class TestNestedFunctionSet(TestCase):
+    def setUp(self):
+        self.fs1 = FunctionSet(
+            a = MockFunction((1, 2)))
+        self.fs2 = FunctionSet(
+            fs1 = self.fs1,
+            b  = MockFunction((3, 4)))
+
+    def test_get_sorted_funcs(self):
+        self.assertItemsEqual([k for (k, v) in self.fs2._get_sorted_funcs()], ('b', 'fs1'))
```
Comment (beam2d) [inline tests/test_function_set.py]: Use `keys()` instead of the list comprehension: `self.fs2._get_sorted_funcs().keys()`

File: tests/test_function_set.py [hunk 10]
```diff
@@ -1,10 +1,63 @@
...
+
+    def test_collect_parameters(self):
+        p_b = np.zeros((3, 4)).astype(np.float32)
+        p_a = np.zeros((1, 2)).astype(np.float32)
+        gp_b = np.ones((3, 4)).astype(np.float32)
+        gp_a = np.ones((1, 2)).astype(np.float32)
+
+        actual = self.fs2.collect_parameters()
```
Comment (beam2d) [inline tests/test_function_set.py]: Check the length of parameter/gradient lists before accessing there elements.

File: tests/test_function_set.py [hunk 11]
```diff
@@ -12,6 +65,9 @@ def setUp(self):
             b = Linear(3, 2)
         )
 
+    def test_get_sorted_funcs(self):
+        self.assertItemsEqual([k for (k, v) in self.fs._get_sorted_funcs()], ('a', 'b'))
```
Comment (beam2d) [inline tests/test_function_set.py]: Use `keys` instead of the list comprehension as above.

File: tests/test_optimizer.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
+from unittest import TestCase
+import numpy as np
+from chainer import cuda, Optimizer
+from chainer.optimizer import _sqnorm
```
Comment (beam2d) [inline tests/test_optimizer.py]: Add a TODO comment to make _sqnorm public API (e.g. in the `utils` subpackage).

File: tests/test_optimizer.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+cuda.init()
+
+class TestOptimizerUtility(TestCase):
+    def setUp(self):
+        # x is an arithmetic progression of length 6
+        # whose common difference is 0.5
```
Comment (beam2d) [inline tests/test_optimizer.py]: Remove this comment. It is described enough by the source code.

**Comment by delta2323** [comment]
@beam2d Currently we do not have any unittest for `Optimizer` class, although derived classes have simple functionality tests. At least I would like to add it.

I am in favor of @beam2d's suggestion that we should create new issues or PRs.

File: tests/test_variable.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+def constant(xs, value):
+    return Constant(value)(*xs)
+
+class TestVariable(TestCase):
+    def setUp(self):
+        self.x = np.random.uniform(-1, 1, 10).astype(np.float32)
+        self.a = np.random.uniform(0.1, 10, 10).astype(np.float32)
+
+    def check_len(self, x, gpu):
```
Comment (beam2d) [inline tests/test_variable.py]: Do we need to pass `x` here?

File: tests/test_variable.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+    def test_len_cpu(self): self.check_len(self.x, False)
+    def test_len_gpu(self): self.check_len(self.x, True)
+
+    def check_backward(self, inputs, intermediates, outputs, retain_grad):
+        for o in outputs:
+            o.backward(retain_grad)
+
+        self.assertFalse(any([x.grad is None for x in inputs]))
+        intermediate_grads = [x.grad is None for x in intermediates]
```
Comment (beam2d) [inline tests/test_variable.py]: This variable name is confusing. Change the name or the definition.

**Comment by beam2d** [comment]
I added comments, so please resolve them. I also found that we should add dedicated unittests for following functions **in future PRs**.
- `Variable.set_creator`, `Variable.unchain_backward`
- Direct usage of `Function.forward` and `Function.backward`
- `Function.unchain`, `Function.to_gpu`, `Function.to_cpu`, `Function.parameters`, `Function.gradients`
- `Split.add_branch`, `Split.backward`
- `FunctionSet.to_gpu`, `FunctionSet.to_cpu`
- `Optimizer.setup`, `Optimizer.init_state`, `Optimizer.zero_grads`, `Optimizer.compute_grads_norm`, `Optimizer.clip_grads`, `Optimizer.weight_decay`, `Optimizer.accumulate_grads`, `Optimizer.update`
- `gradient_check.numerical_grad`
- Functions in `cuda` package
- Functions in `utils` package

File: tests/test_optimizer.py (hunk 13 above)
Comment (delta2323) [inline tests/test_optimizer.py]: I think this TODO comment should be placed where `_sqnorm` is defined, rather than in test code.

File: tests/test_optimizer.py (hunk 12 above)
Comment (delta2323) [inline tests/test_optimizer.py]: I think this TODO comment should be placed where `_sqnorm` is defined, rather than in test code.

File: tests/test_function_set.py (hunk 11 above)
Comment (delta2323) [inline tests/test_function_set.py]: The type of return value of `Function._get_sorted_funcs()` is an array of tuple(especially pair). So do you intended to write `tuple(self.fs._get_sorted_funcs()).keys()` instead?

Comment (beam2d) [inline tests/test_function_set.py]: Ahh, OK, I forgot that. Then the code above is fine. Ignore my comments.

**Comment by beam2d** [comment]
It seems test_ada_delta.py often fails. Setting `eps=1e-5` for AdaDelta resolves this problem in my environment.

**Comment by beam2d** [comment]
Would you adapt to another test-related PR #54 which is merged just now?

**Comment by beam2d** [comment]
Import `chainer.testing.attr` to use the decorators.

**Comment by beam2d** [comment]
[LGTM](http://lgtm.in/i/Ig1FwdbOp)

**Comment by beam2d** [comment]
Thanks a lot!

**Comment by delta2323** [comment]
Thank you very much! I will post a issue on unittests for each module later.
